packages/webpack/lib/middlewares/render.js

```diff
@@ -4,61 +4,70 @@ const { join } = require('path');
+const EnhancedPromise = require('eprom');
 
-  let middleware = null;
+const extractErrors = (stats) => {
+  const { errors } = stats.toJson();
+  const details = errors.map((error) =>
+    error.replace(/\033?\[[0-9]{1,2}m/g, '')
+  );
+  return details.join('\n');
+};
 
-  const enhancedPromise = new EnhancedPromise((resolve, reject, reset) => {
-    const handleCompilation = (compileError, stats) => {
+const getBuildPromise = (webpackConfig) => {
+  const compiler = webpack(webpackConfig);
+  compiler.outputFileSystem = new MemoryFS();
+  sourceMapSupport.install({ environment: 'node', hookRequire: true });
+  return new Promise((resolve, reject) => {
+    compiler.run((compileError, stats) => {
       if (compileError) {
         reject(compileError);
       } else if (stats.hasErrors()) {
-        const { errors } = stats.toJson();
-        const details = errors.map((error) =>
-          error.replace(/\033?\[[0-9]{1,2}m/g, '')
-        );
-        reject(new Error(`Can't compile:\n${details.join('\n')}`));
+        reject(new Error(`Can't compile:\n${extractErrors(stats)}`));
       } else {
         const { path, filename } = webpackConfig.output;
         const filePath = join(path, filename);
-        if (watch) {
-          if (middleware) {
-            process.kill(process.pid, 'SIGUSR2');
-            resolve(middleware);
-          } else {
-            try {
-              middleware = require(filePath).default;
-              resolve(middleware);
-            } catch (moduleError) {
-              reject(moduleError);
-            }
-          }
+        const { outputFileSystem: fs } = compiler;
+        fs.readFile(filePath, 'utf-8', (readError, fileContents) => {
+          if (readError) return reject(readError);
+          resolve(requireFromString(fileContents, filePath));
+        });
+      }
+    });
+  });
+};
+
+const getWatchPromise = (webpackConfig) => {
+  const compiler = webpack(webpackConfig);
+  let middleware = null;
+  return new EnhancedPromise((resolve, reject, reset) => {
+    compiler.hooks.watchRun.tap('RenderMiddleware', () => reset());
+    compiler.watch(webpackConfig.watchOptions, (compileError, stats) => {
+      if (compileError) {
+        reject(compileError);
+      } else if (stats.hasErrors()) {
+        reject(new Error(`Can't compile:\n${extractErrors(stats)}`));
+      } else {
+        if (middleware) {
+          process.kill(process.pid, 'SIGUSR2');
         } else {
-          const { outputFileSystem: fs } = compiler;
-          fs.readFile(filePath, 'utf-8', (readError, fileContents) => {
-            if (readError) return reject(readError);
-            resolve(requireFromString(fileContents, filePath).default);
-          });
+          const { path, filename } = webpackConfig.output;
+          const filePath = join(path, filename);
+          middleware = require(filePath);
         }
+        process.nextTick(() => resolve(middleware));
```

Comment:
I just tested this and it turns out that `process.nextTick()` is actually not required :see_no_evil: 

```suggestion
        resolve(middleware);
```